File: frontends/android/app/src/main/java/com/example/animastor/repository/AuthStore.kt
package com.example.animastor.repository

import com.example.animastor.network.RetrofitClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import retrofit2.HttpException

/**
 * Authentication state — web parity with frontends/app state/authStore.ts.
 *
 * The server-side cookie is the only source of truth (HttpOnly `animastor_sid`
 * session + `animastor_gid` guest); this store only mirrors GET /auth/me and
 * forwards login/register/logout to the /api/v1/auth endpoints. No tokens are
 * ever stored or handled here.
 */
object AuthStore {

    data class AuthState(
        val authenticated: Boolean = false,
        /** "user" | "guest" | "none" — mirrors /auth/me identity. */
        val identity: String = "none",
        val user: AuthUser? = null,
        val workspace: AuthWorkspace? = null
    )

    private val _state = MutableStateFlow(AuthState())
    val state: StateFlow<AuthState> = _state.asStateFlow()

    private val _busy = MutableStateFlow(false)
    val busy: StateFlow<Boolean> = _busy.asStateFlow()

    /** Refresh identity from the server. Safe on network failure. */
    suspend fun refresh(): AuthState {
        val me = try {
            RetrofitClient.api.me()
        } catch (_: Exception) {
            null
        }
        val next = me?.let {
            AuthState(
                authenticated = it.authenticated,
                identity = it.identity ?: if (it.authenticated) "user" else "none",
                user = it.user,
                workspace = it.workspace
            )
        } ?: AuthState()
        _state.value = next
        return next
    }

    suspend fun login(username: String, password: String) {
        _busy.value = true
        try {
            val me = RetrofitClient.api.login(AuthLoginRequest(username, password))
            _state.value = AuthState(true, me.identity ?: "user", me.user, me.workspace)
        } finally {
            _busy.value = false
        }
    }

    suspend fun register(username: String, password: String, email: String?) {
        _busy.value = true
        try {
            val me = RetrofitClient.api.register(AuthRegisterRequest(username, password, email))
            _state.value = AuthState(true, me.identity ?: "user", me.user, me.workspace)
        } finally {
            _busy.value = false
        }
    }

    /** Logout is idempotent server-side; the cookie jar clears regardless. */
    suspend fun logout() {
        _busy.value = true
        try {
            runCatching { RetrofitClient.api.logout() }
        } finally {
            com.example.animastor.AnimastorApp.authCookies?.clear()
            _state.value = AuthState()
            _busy.value = false
        }
    }
}

File: frontends/app/src/state/authStore.ts
// Authentication state (Account & Workspace MVP).
// Server-side session cookie is the source of truth; localStorage is never
// used for auth. The store only mirrors /auth/me.
import { signal } from '@preact/signals';
import { getJson, postJson } from '../api/client';
import { stashBookSessionForUser, restoreStashedBookSessionForUser } from './generateStore';

export interface AuthUser { id: string; username: string; display_name?: string | null; role?: string; }
export interface AuthWorkspace { id: string; name: string; type: string; }
export interface AuthMe { authenticated: boolean; user: AuthUser | null; workspace: AuthWorkspace | null; }

export const authMe = signal<AuthMe>({ authenticated: false, user: null, workspace: null });
export const authLoading = signal(false);
export const authError = signal<string | null>(null);

export async function fetchMe(): Promise<AuthMe> {
  try {
    const me = await getJson<AuthMe>('/auth/me');
    authMe.value = me;
    authError.value = null;
    return me;
  } catch {
    // Unauthenticated / server unreachable — reflect as anonymous.
    authMe.value = { authenticated: false, user: null, workspace: null };
    return authMe.value;
  }
}

export async function login(username: string, password: string): Promise<AuthMe> {
  authLoading.value = true;
  authError.value = null;
  try {
    const me = await postJson<AuthMe>('/auth/login', { username, password });
    authMe.value = me;
    // Current book session re-attach: if this user had a book open when they
    // last logged out, bring it back now (validated by restoreBookSession()).
    if (me?.user?.id) restoreStashedBookSessionForUser(me.user.id);
    return me;
  } catch (e) {
    authError.value = e instanceof Error ? e.message : 'login failed';
    throw e;
  } finally {
    authLoading.value = false;
  }
}

export async function register(username: string, password: string, email?: string): Promise<AuthMe> {
  authLoading.value = true;
  authError.value = null;
  try {
    const me = await postJson<AuthMe>('/auth/register', { username, password, email });
    authMe.value = me;
    return me;
  } catch (e) {
    authError.value = e instanceof Error ? e.message : 'registration failed';
    throw e;
  } finally {
    authLoading.value = false;
  }
}

export async function logout(): Promise<void> {
  authError.value = null;
  // Capture the identity BEFORE clearing it: the open book session is stashed
  // under the user's own key so the same user gets their book back on the
  // next login (the book itself stays owned by them in the DB).
  const userId = authMe.value.user?.id ?? null;
  try {
    await postJson<unknown>('/auth/logout', {});
  } catch { /* logout is idempotent server-side; cookie clear best-effort */ }
  // Current book session isolation: the live session (localStorage
  // 'animastor:currentBook' + open-book signals) must NOT survive into the
  // anonymous/guest context — otherwise a browser refresh would re-open the
  // previous user's book. Ownership is unchanged; the session is merely
  // stashed per-user by stashBookSessionForUser().
  stashBookSessionForUser(userId);
  authMe.value = { authenticated: false, user: null, workspace: null };
}
